### DISCUSSION TITLE
How would I build a multi-platform build that I can upload to an internal registry?

### DISCUSSION THREAD

**Discussion by fatso83** [ibm-messaging/mq-container#574] [Q&A]
I have [previously](https://github.com/ibm-messaging/mq-container/issues/562#issuecomment-2134635123) built a developer image on my computer to test out MQ code locally, but I could never commit an update to our `docker-compose.yml`, as that would fail on their machines. 

So I want to upload a version that works on both x86 and ARM computers, but it's not immediately clear how. If I follow the docs and run `make build-devserver` I end up with a local image that is tagged with `-arm64`. I assume that would I want is to have it build _two_ images, one for `linux/amd64` and one for `linux/arm64`, where the name is the same, so that the same name could be used in the Docker file ([multiplatform builds](https://docs.docker.com/build/building/multi-platform/)).

I would then `docker push` that image to our internal registry so that it could be used by all devs on our team.

Is there any way of achieving this today?

**Comment by arthurbarr** [comment]
Yes, we do this for all the official images.  You need to create the separate images, and then create a "manifest" which ties those images together.  See [podman manifest](https://docs.podman.io/en/latest/markdown/podman-manifest.1.html) for more information.

**Reply by fatso83** [reply]
Thank you. 

Wrt licenses, etc I just noted this bit in the README

> Note: The IBM MQ Advanced for Developers license does not permit further distribution and the terms restrict usage to a developer machine.

Will building the images and uploading it to an internal (non-public) registry constitute "further distribution"?

**Reply by fatso83** [reply]
Somehow I think my struggles here are more like a symptom. Any idea why IBM is not just publishing both platforms? I mean, all the bits are in place, there is a developer build for x86 already and it builds fine on mac, so it just seems like a nuisance that could so easily be remedied.

**Reply by arthurbarr** [reply]
It's just down to pipeline limitations.  The same pipeline is used for MQ Advanced for Developers images as the production images, and it expects three architectures — if we change this, it will sometimes be four architectures.  That all needs to be tested, and our internal pipeline is undergoing a lot of other changes at the moment.  It's not a particularly good reason, but that's the reason.

**Reply by fatso83** [reply]
Thank you. Totally understandable. With regards to the licensing question, would uploading to an internal tracker be permitted? I am not sure if this is regarded as distribution in the sense of the readme and would not like to get lawsuits by not being mindful 😅

**Reply by SoundaryaJayakumar** [reply]
Hi! On the licensing question, it depends on the purpose. If the purpose of building the images and uploading it to an internal (non-public) registry is to use it at a later date yourself, that would not constitute distribution. And I believe that's what you're referring to. It would be distribution only If other developers can access that copy and make a download from it so to individually use it then that wouldn't be allowed.

**Reply by fatso83** [reply]
@SoundaryaJayakumar Thank you for the reply!

It would actually be done with the intent of having other developers on my company's team use it, so they would not have to go through the process of manually downloading it, building it and figuring out all the details. That seems to be in conflict with the licensing terms, if I understand you right.

Given that, I'll refrain from uploading the built artifact (unless I'm mistaken about something), but it's really a bit of a hassle that we as a commercial consumer of IBM services (such as the production MQ server) really should not have to go through when trying to develop integrations, so hopefully there might be an added clause on how this does not apply to paying customers or something in the future 🙏
